### PR TITLE
XXE Challenge 1

### LINKED ISSUE #547: XXE Challenge 1
This should be harder than lesson created in #334

### PR DESCRIPTION
**Pull request by ismisepaul**
First XXE Challenge where the app sends JSON to an endpoint and you need to find the old XML endpoint.

closes #556

### CODE REVIEW

File: src/main/java/servlets/module/challenge/XxeChallenge1.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                    if (Validate.validateTokens(tokenCookie, tokenHeader)) {
+                        InputStream json = request.getInputStream();
+                        String emailAddr = readJson(json);
+                        log.debug("Email Addr: " + emailAddr);
+
+                        String htmlOutput = new String();
+
+                        if (emailAddr == null) {
+                            htmlOutput += "<p>" + bundle.getString("response.blank.email") + "</p>";
+                            out.write(htmlOutput + emailAddr);
```
Comment (markdenihan) [inline src/main/java/servlets/module/challenge/XxeChallenge1.java:96]: Should emailAddr be encoded for HTML output?

File: src/main/java/servlets/module/challenge/XxeChallenge1.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                        } else if (Validate.isValidEmailAddress(emailAddr)) {
+                            log.debug("User Submitted - " + emailAddr);
+
+                            htmlOutput += "<p>" + bundle.getString("response.success.reset") + ": " + emailAddr
```
Comment (markdenihan) [inline src/main/java/servlets/module/challenge/XxeChallenge1.java:100]: hould emailAddr be encoded for HTML output?

File: src/main/java/servlets/module/challenge/XxeChallenge1.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                                    + " has been reset</p>";
+                            out.write(htmlOutput);
+                        } else {
+                            htmlOutput += "<p>" + bundle.getString("response.invalid.email") + ": "
+                                    + emailAddr + "</p>";
```
Comment (markdenihan) [inline src/main/java/servlets/module/challenge/XxeChallenge1.java:105]: hould emailAddr be encoded for HTML output?

**Review by markdenihan — CHANGES_REQUESTED** [review]

File: src/main/java/servlets/module/challenge/XxeChallenge1.java (hunk 1 above)
Comment (ismisepaul) [reply src/main/java/servlets/module/challenge/XxeChallenge1.java:96]: is done at line ```89: emailAddr = Encode.forHtml(emailAddr);```

File: src/main/java/servlets/module/challenge/XxeChallenge1.java (hunk 2 above)
Comment (ismisepaul) [reply src/main/java/servlets/module/challenge/XxeChallenge1.java:100]: is done at line ```89: emailAddr = Encode.forHtml(emailAddr);```

File: src/main/java/servlets/module/challenge/XxeChallenge1.java (hunk 3 above)
Comment (ismisepaul) [reply src/main/java/servlets/module/challenge/XxeChallenge1.java:105]: is done at line ```89: emailAddr = Encode.forHtml(emailAddr);```

**Review by markdenihan — APPROVED** [review]
+1 for emoji use

**Pull request merged**